# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

導入數據分析需要的庫並通過Pandas的read_csv函數，將原始數據文件"e_commerce.csv"數據內容解析為DataFrame並賦值給變量original_data

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

In [4]:
original_data.sample(10)

InvoiceNo StockCode                      Description  Quantity  \
247641    558774     22211           WOOD STAMP SET FLOWERS         1   
338695    566562     22141   CHRISTMAS CRAFT TREE TOP ANGEL         5   
389271    570466     22951  60 CAKE CASES DOLLY GIRL DESIGN         1   
119866    546634     23182   TOILET SIGN OCCUPIED OR VACANT         3   
436305    574097     23474   WOODLAND SMALL BLUE FELT HEART        12   
220456    556193     22489   PACK OF 12 TRADITIONAL CRAYONS        24   
228945    557002     22726       ALARM CLOCK BAKELIKE GREEN        48   
149157    549278     22151        PLACE SETTING WHITE HEART        96   
173097    551653     22730       ALARM CLOCK BAKELIKE IVORY         4   
327814    565728     22086  PAPER CHAIN KIT 50'S CHRISTMAS         24   

             InvoiceDate  UnitPrice  CustomerID         Country  
247641    7/4/2011 10:17       0.83         NaN  United Kingdom  
338695   9/13/2011 12:18       2.10     17365.0  United Kingdom  
389271  10/10/2011 16:03       0.55     16791.0  United Kingdom  
119866   3/15/2011 12:10       0.83     14527.0  United Kingdom  
436305    11/3/2011 9:56       0.83     12471.0         Germany  
220456    6/9/2011 12:04       0.42     14132.0  United Kingdom  
228945   6/16/2011 10:49       3.75     16672.0  United Kingdom  
149157    4/7/2011 13:16       0.42     16255.0  United Kingdom  
173097    5/3/2011 12:03       3.75     13282.0  United Kingdom  
327814    9/6/2011 12:00       2.95     17716.0  United Kingdom

評估數據乾淨度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


輸出結果來看，數據共541909條，Description & CustomerID 變數存在缺失值，另外InvoiceDate應屬於日期，CustomerID應屬於字符串，
應進行上列數據格式轉換。

**評估缺失數據

了解'Description' 存在缺失值並根據條件提取出缺失觀察值

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

發現1454條交易數據缺失Description交易數據，UnitPrice可能為0，驗證是否Description變量缺失且UnitPritce不為0的數據。

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0) ]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

解析結果數量為0條，確認Description值數據同時不具有效UnitPrice值，
Description為產品名稱，UnitPrice為產品單價，都是後續進行交易分系重要數值，同時缺失無效造成後續無效分析，建議進行刪除。

優先提取分析CustomerID存在缺失值，因此根據條件提除缺失觀察值

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID 數據缺失，屬於客戶編號，但135080條有其他數據，輸出結果看來仍屬於有效，因此保留此變數為空的觀察值。

**評估不一致數據

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


***調查是否有重複或無效數據

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

輸出結果Quantity & UnitPrice 存在負數值，對後續數據分析會產生影響的，因此先提取出Quantity負數觀察值進一步評估。

需要驗證猜想，增加篩選條件，檢查Quantity變量為負數且InvoiceNo 非C開頭的觀察值。

In [11]:
original_data[(original_data["Quantity"] <0) & (original_data["InvoiceNo"].str[0] !="C") ]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

需要驗證猜想，增加篩選條件，檢查Quantity變量為負數且InvoiceNo 非C開頭的觀察值，多加一條篩選UnitPrice 條件進行驗證

In [13]:
original_data[(original_data["Quantity"] <0) & (original_data["InvoiceNo"].str[0] !="C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

## 清理数据

Quantity 變量負數滿足下列條件之一:
1. InvoiceNo C開頭表示取消訂單
2. UnitPrice 為0 說明單價0元
這些交易均不是有效成交數據，不在後續分析範圍內，依序清除Quantity負數觀察值，再篩選出UnitPrice為負數觀察值。

In [15]:
original_data[original_data["UnitPrice"]<0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

主要工作
1. InvoiceDate 變量轉換日期時間
2. CustomerID 變量轉換字符串
3. Description 變量缺失觀察值刪除
4. Country變量USA替換United States
5. Country變量UK、U.K、替換United Kingdom
6. Quantity變量負觀察值刪除
7. UnitPrice變量負數觀察值刪除

創建新的變量cleand_data,讓original_data複製出的副本清理後運用在cleand_data上。

In [16]:
cleand_data = original_data.copy()
cleand_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

進行日期轉換

In [17]:
cleand_data["InvoiceDate"] = pd.to_datetime(cleand_data["InvoiceDate"])
cleand_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

客戶轉換字符串

In [18]:
cleand_data["CustomerID"] = cleand_data["CustomerID"].astype(str)
cleand_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把結尾客戶ID.0刪除

In [20]:
cleand_data["CustomerID"] = cleand_data["CustomerID"].str.slice(0,-2)
cleand_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

Description 變量缺失觀察值刪除並查看刪除後空缺值個數和

In [22]:
cleand_data.dropna(subset=["Description"], inplace=True)

In [23]:
cleand_data["Description"].isnull().sum()

np.int64(0)

Country 變量USA轉換United States ，檢查轉換後USA變量個數

In [26]:
cleand_data["Country"] = cleand_data["Country"].replace({"USA": "United States"})

In [27]:
len(cleand_data[cleand_data["Country"] == "USA"])

0

變更UK轉換

In [28]:
cleand_data["Country"] = cleand_data["Country"].replace({"UK": "United Kingdom","U.K.": "United Kingdom"})

In [31]:
print(len(cleand_data[cleand_data["Country"] == "UK"] ))
print(len(cleand_data[cleand_data["Country"] == "U.K."] ))

0
0


Qunatity 負值清除後清除UnitPrice負值刪除

In [33]:
cleand_data = cleand_data[cleand_data["Quantity"] >=0]

In [43]:
len(cleand_data[cleand_data["Quantity"] < 0 ])

0

In [40]:
cleand_data = cleand_data[cleand_data["UnitPrice"] >=0]

In [42]:
len(cleand_data[cleand_data["UnitPrice"] < 0 ])

0

## 保存清理后的数据

In [44]:
cleand_data.to_csv("e_commerce_cleand.csv", index=False)

In [45]:
pd.read_csv("e_commerce_cleand.csv").head()

C:\Users\PC\AppData\Local\Temp\ipykernel_17380\3469763782.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleand.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom